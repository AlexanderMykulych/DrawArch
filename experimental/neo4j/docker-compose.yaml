version: '3'

services:
  neo4j:
    image: neo4j:latest
    ports:
      - "7474:7474"
      - "7687:7687"
    environment:
      # - NEO4J_dbms_security_procedures_unrestricted=apoc.*,algo.*
      # - NEO4J_dbms_security_procedures_whitelist=apoc.*,algo.*
      - NEO4J_apoc_import_file_enabled=true
      - NEO4J_dbms_shell_enabled=true
      - NEO4JLABS_PLUGINS=["apoc"]
      - NEO4J_AUTH=neo4j/test
    volumes:
      - $HOME/neo4j/plugins:/plugins
      - $HOME/neo4j/data:/data
      - $HOME/neo4j/import:/import
      - $HOME/neo4j/neo4j/logs:/logs