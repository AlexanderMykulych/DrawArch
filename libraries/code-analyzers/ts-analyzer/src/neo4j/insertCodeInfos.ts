import type { Driver } from "neo4j-driver";
import { CodeInfo, CodeInfoType, Entity, Relationship } from "../analysis/types";
import { groupBy } from 'coya-util'
import { progress } from "../progress/progress";

export function getInsertCodeInfosFn(driver: Driver, database: string) {

  async function insertCodeInfos(codeInfos: CodeInfo[]) {
    const session = driver.session({ database })

    await session.writeTransaction(tx => {
      Object.entries(
        groupBy(
          codeInfos
            .filter((x): x is Entity => x.type === CodeInfoType.Entity),
          (item) => item.entityType,
        )
      ).map(([type, entities]) =>
        tx.run(`CALL apoc.create.nodes(["${type}"], $entities)`, { entities })
      )


    })
    await session.writeTransaction(tx => {
      codeInfos
        .filter((x): x is Relationship => x.type === CodeInfoType.Relationship)
        .map(r => ({
          query: `
            MATCH (e1 { id: $from})
            MATCH (e2 { id: $to})
            CALL apoc.create.relationship(e1, "Relation", $relation, e2)
            YIELD rel
            RETURN rel
            `,
          params: {
            relation: r,
            from: r.from,
            to: r.to,
          },
        }))
        .map(x => tx.run(x.query, x.params))
    })
    await session.close()
  }

  return progress('neo4j. insert', insertCodeInfos)
}