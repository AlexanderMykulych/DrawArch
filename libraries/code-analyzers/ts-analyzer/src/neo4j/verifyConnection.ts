import type { Driver } from 'neo4j-driver'
import { progress } from '../progress/progress'

export function getVerifyConnectionFn(driver: Driver) {
  function verifyConnection() {
    return driver.verifyConnectivity()
  }
  return progress('neo4j. verify', verifyConnection)
}
