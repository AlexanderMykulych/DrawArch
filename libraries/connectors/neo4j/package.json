{
  "name": "coya-neo4j",
  "type": "module",
  "version": "1.0.0",
  "description": "",
  "main": "./dist/index.js",
  "types": "./index.ts",
  "scripts": {
    "build": "vite build"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "vite": "^3.1.0",
    "vite-node": "~0.23.1",
    "axios": "~1.3.2",
    "flatted": "~3.2.7",
    "neo4j-driver": "~5.5.0",
    "coya-connectors-shared": "~1.0.0",
    "@antfu/eslint-config": "~0.35.2",
    "ws": "~8.12.0"
  }
}
