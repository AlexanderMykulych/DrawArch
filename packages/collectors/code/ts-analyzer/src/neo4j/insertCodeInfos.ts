import type { Driver } from "neo4j-driver";
import { CodeInfo, CodeInfoType, Entity, Relationship } from "../analysis/types";
import { groupBy } from 'coya-util'
import { progress } from "../progress/progress";

export function getInsertCodeInfosFn(driver: Driver, database: string) {

  async function insertCodeInfos(codeInfos: CodeInfo[]) {
    const session = driver.session({ database })

    await session.writeTransaction(tx => {
      Object.entries(
        groupBy(
          codeInfos
            .filter((x): x is Entity => x.type === CodeInfoType.Entity),
          (item) => item.entityType,
        )
      ).map(([type, entities]) =>
        tx.run(`CALL apoc.create.nodes(["${type}"], $entities)`, { entities })
      )
    })
    await session.writeTransaction(tx => {
      const rows = codeInfos
        .filter((x): x is Relationship => x.type === CodeInfoType.Relationship)

      tx.run(`
          UNWIND $batch as row
          MATCH (from { id: row.from})
          MATCH (to { id: row.to})
          CALL apoc.create.relationship(from, "Relation", row, to)
          YIELD rel
          RETURN count(*)
          `,
        {
          batch: rows,
        },
      )
    })
    await session.close()
  }

  return progress('neo4j. insert', insertCodeInfos)
}