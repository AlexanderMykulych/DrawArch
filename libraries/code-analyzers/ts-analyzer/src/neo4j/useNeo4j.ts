import neo4j from 'neo4j-driver'
import type { Driver } from 'neo4j-driver'
import { groupBy } from 'coya-util'
import { CodeInfo, CodeInfoType, Entity, Relationship } from '../analysis/types'

let driver: Driver
const database = 'neo4j'
export function useNeo4j() {

  driver ??= neo4j.driver('neo4j://localhost:7687', neo4j.auth.basic('neo4j', 'test'))

  return {
    verifyConnection() {
      return driver.verifyConnectivity()
    },
    async init() {
      const session = driver.session({ database })
      await session.writeTransaction(tx => {
        tx.run(`MATCH (n) DETACH DELETE n`)
      })

      await session.close()
    },
    async insertCodeInfos(codeInfos: CodeInfo[]) {
      const session = driver.session({ database })

      await session.writeTransaction(tx => {
        Object.entries(
          groupBy(
            codeInfos
              .filter((x): x is Entity => x.type === CodeInfoType.Entity),
            (item) => item.entityType,
          )
        ).map(([type, entities]) =>
          tx.run(`CALL apoc.create.nodes(["${type}"], $entities)`, { entities })
        )


      })
      await session.writeTransaction(tx => {
        codeInfos
          .filter((x): x is Relationship => x.type === CodeInfoType.Relationship)
          .map(r => ({
            query: `
            MATCH (e1 { id: "${r.from}"})
            MATCH (e2 { id: "${r.to}"})
            CALL apoc.create.relationship(e1, "Relation", $relation, e2)
            YIELD rel
            RETURN rel
            `,
            params: {
              relation: r,
              from: r.from,
              to: r.to,
            },
          }))
          .map(x => tx.run(x.query, x.params))
      })
      await session.close()
    }
  }
}
