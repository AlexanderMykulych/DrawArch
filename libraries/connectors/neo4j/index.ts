import type { Relation } from 'coya-connectors-shared'
import neo4j from 'neo4j-driver'

export function getNeo4j() {
  const database = neo4j.driver('neo4j+s://aa84864b.databases.neo4j.io', neo4j.auth.basic('neo4j', 'BHK4W_phDl5-U9TVZ7iFic3Io4pm-CrPYB4cWoAnWhc'))
  return {
    async insert<T extends Object>(label: string, items: T[]) {
      const session = database.session()

      const query = `UNWIND $items AS node CREATE (n:${label}) SET n = node`
      try {
        await session.run(query, { items })
      }
      finally {
        await session.close()
      }
    },
    async insertRelations(items: Relation[]) {
      const session = database.session()

      const query = `
      UNWIND $items AS relation
      CALL apoc.merge.node([relation.fromNode], { id: relation.from }) YIELD node as n1
      CALL apoc.merge.node([relation.toNode], { id: relation.to }) YIELD node as n2
      CALL apoc.merge.relationship(n1, relation.type, relation, {}, n2) YIELD rel
      RETURN n1, n2, rel
      `
      try {
        const chunks = chunkArray(items, 200)

        let chunkIndex = 0

        for await (const chunk of chunks) {
          await session.run(query, { items: chunk })
          console.log(`chunk ${chunkIndex++} inserted`)
        }
      }
      finally {
        await session.close()
      }
    },
    async clearDb() {
      const session = database.session()

      await session.run('MATCH (n) DETACH DELETE n')
    },
    async query(query: string, params?: any) {
      const session = database.session()

      return await session.run(query, params)
    },
  }
}

function chunkArray<T>(array: T[], chunkSize: number) {
  const result = []
  for (let i = 0; i < array.length; i += chunkSize) {
    result.push(array.slice(i, i + chunkSize))
  }
  return result
}
