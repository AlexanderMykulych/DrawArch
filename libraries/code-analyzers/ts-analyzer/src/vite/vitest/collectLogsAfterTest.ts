import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}