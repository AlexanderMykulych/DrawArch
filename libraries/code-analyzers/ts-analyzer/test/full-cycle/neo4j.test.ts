import { expect, test } from "vitest";
import { useNeo4j } from "../../src/neo4j/useNeo4j";


test('neo4j is ready to work', async () => {
  const db = useNeo4j()

  await db.init()

  expect(await db.verifyConnection()).toMatchSnapshot()
}, 200)