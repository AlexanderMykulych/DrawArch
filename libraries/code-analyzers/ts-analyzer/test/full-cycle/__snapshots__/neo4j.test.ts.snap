// Vitest Snapshot v1

exports[`neo4j is ready to work 1`] = `
{
  "address": "localhost:7687",
  "version": "Neo4j/4.4.7",
}
`;
